### PR TITLE
Add UDPs to dictionary filtering

### PR DESCRIPTION
**Pull request by pwoody**
@robert3005

### CODE REVIEW

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java [hunk 1]
```diff
@@ -358,6 +364,39 @@ public void testOr() throws Exception {
...
+  public void testInverseUdp() throws Exception {
+    InInt32UDP udp = new InInt32UDP(ImmutableSet.of(42));
+    InInt32UDP udp1 = new InInt32UDP(ImmutableSet.of(205));
+    FilterPredicate inverse =
+      LogicalInverseRewriter.rewrite(FilterApi.not(FilterApi.userDefined(intColumn("int32_field"), udp)));
+    FilterPredicate inverse1 =
+      LogicalInverseRewriter.rewrite(FilterApi.not(FilterApi.userDefined(intColumn("int32_field"), udp1)));
+
+    assertFalse("Should not drop block for inverse of non-matching UDP",
+      canDrop(inverse
```
Comment (robert3005) [inline parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java]: same line

**Comment by robert3005** [comment]
👍

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java [hunk 2]
```diff
@@ -464,6 +508,50 @@ public void testGtEqMissingColumn() throws Exception {
...
+    }
+
+    @Override
+    public boolean keep(Integer value) {
+      return ints.contains(value);
+    }
+
+    @Override
+    public boolean canDrop(Statistics<Integer> statistics) {
+      return false;
```
Comment (ash211) [inline parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java]: is this a stub value?  We ought to be able to check the stats against the integer set to more intelligently filter the file out I'd think?

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java [hunk 3]
```diff
@@ -464,6 +508,50 @@ public void testGtEqMissingColumn() throws Exception {
...
+    }
+
+    @Override
+    public boolean inverseCanDrop(Statistics<Integer> statistics) {
+      return false;
```
Comment (ash211) [inline parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java]: stub value here too?

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java (hunk 2 above)
Comment (robert3005) [inline parquet-hadoop/src/test/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilterTest.java]: This is just for test. canDrop, inverseCanDrop are used for statistics filtering while keep is used for row level and dictionary filtering.

**Pull request merged**